### PR TITLE
fix(dashboard): unwrap tool-result JSON in the Output tab (#5778)

### PR DESCRIPTION
**Pull request by blamechris**
## Problem

The Output tab (the simulated terminal view of session output) rendered the **raw JSON envelope** of a tool result, e.g.:

```
{"stdout":"total 0\ndrwxr-xr-x@ 2 blamechris staff 64 Jun 14 ...","stderr":"",...}
```

instead of the clean stdout text rendered as terminal lines. For shell-style tools (Bash, etc.) the tool-result string is itself a JSON-stringified `{ stdout, stderr, ... }` envelope, and `handleToolResult` forwarded that string verbatim to `appendTerminalData`.

## Fix

New pure helper `packages/dashboard/src/lib/tool-result-text.ts` → `unwrapToolResultText()`:

- **plain strings** pass through unchanged
- **`{ stdout, stderr }` envelopes** → render the streams (stdout, with stderr appended); empty streams omitted
- **any other shape** (non-envelope object, JSON array, invalid JSON, non-string) → falls back to the original string, never throws

Wired into the terminal-write side effect in `handleToolResult` (message-handler.ts). The chat-side `ToolBubble` and `result.resultText`/`toolResult` history are untouched — only the Output tab's terminal preview is affected.

## Tests

`tool-result-text.test.ts` covers plain text, stdout-only, stdout+stderr, stderr-only, non-envelope objects, invalid JSON, brace-leading plain text, and arrays.

## Verification notes

This worktree has no node_modules (workspace hoist), so I could not run `tsc`/`vitest` locally — central verification needed. The helper is pure and self-contained; risk is low. The only behavioral surface beyond the helper is the single call site on the terminal preview path.

Closes #5778

### CODE REVIEW

File: packages/dashboard/src/lib/tool-result-text.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+function streamsToText(obj: Record<string, unknown>): string | null {
+  const hasStdout = typeof obj.stdout === 'string'
+  const hasStderr = typeof obj.stderr === 'string'
+  if (!hasStdout && !hasStderr) return null
+  const parts: string[] = []
+  if (hasStdout && obj.stdout !== '') parts.push(obj.stdout as string)
+  if (hasStderr && obj.stderr !== '') parts.push(obj.stderr as string)
+  return parts.join('\n')
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/dashboard/src/lib/tool-result-text.ts]: **FIX** — `6b17f8c3f`. Replaced the `parts.join('\n')` with a terminal-accurate concat: only insert a separating newline when stdout doesn't already end with one, so trailing-newline shell output no longer gains a spurious blank line before stderr.

```diff
-  const parts: string[] = []
-  if (hasStdout && obj.stdout !== '') parts.push(obj.stdout as string)
-  if (hasStderr && obj.stderr !== '') parts.push(obj.stderr as string)
-  return parts.join('\n')
+  const stdout = hasStdout && obj.stdout !== '' ? (obj.stdout as string) : ''
+  const stderr = hasStderr && obj.stderr !== '' ? (obj.stderr as string) : ''
+  if (!stdout) return stderr
+  if (!stderr) return stdout
+  const sep = stdout.endsWith('\n') ? '' : '\n'
+  return stdout + sep + stderr
```

Covered by a new unit test (`does not double a trailing newline when stdout already ends in one`).

File: packages/dashboard/src/lib/tool-result-text.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+/**
+ * Unwrap a tool-result string for terminal-style display. Returns the
+ * human-readable text; never throws.
+ */
+export function unwrapToolResultText(resultText: string): string {
+  if (typeof resultText !== 'string') return String(resultText ?? '')
+
+  const trimmed = resultText.trim()
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/dashboard/src/lib/tool-result-text.ts]: **FIX** — `6b17f8c3f`. Widened the signature to `unknown` so the runtime guard and the module docs match the contract (callers forward loosely-typed wire values). After the `typeof !== 'string'` guard TS narrows back to `string`, so the rest of the body is unchanged.

```diff
-export function unwrapToolResultText(resultText: string): string {
+export function unwrapToolResultText(resultText: unknown): string {
   if (typeof resultText !== 'string') return String(resultText ?? '')
```

Added a unit test asserting non-string inputs (`null`/`undefined`/`number`) coerce to a safe string.

File: packages/dashboard/src/store/message-handler.ts [hunk 3]
```diff
@@ -2051,10 +2052,13 @@ function handleToolResult(msg: Record<string, unknown>, get: MsgGet, set: MsgSet
...
+  // stdout/stderr text rather than the raw `{"stdout":...}` JSON.
   if (result.resultText) {
-    const preview = result.resultText.length > 500
-      ? result.resultText.slice(0, 500) + '...'
-      : result.resultText;
+    const text = unwrapToolResultText(result.resultText);
+    const preview = text.length > 500
+      ? text.slice(0, 500) + '...'
+      : text;
     get().appendTerminalData(`\x1b[2m${preview}\x1b[0m\r\n`);
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/dashboard/src/store/message-handler.ts:2062]: **FIX** — `6b17f8c3f`. Added an integration-style block to `message-handler.test.ts` (`Output-tab tool_result unwrap (#5778)`) that drives a `tool_result` through `handleMessage` and asserts the terminal-preview side effect:

- JSON envelope `{"stdout":"total 0\ndrwxr-xr-x","stderr":"",...}` → `appendTerminalData` receives the unwrapped stdout, and the write contains no `"stdout"` / `{` (no raw JSON leak).
- plain-string result passes through unchanged.

This guards the `handleToolResult` wiring, not just the helper.

File: packages/dashboard/src/lib/tool-result-text.test.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      stderr: '',
+      interrupted: false,
+    })
+    expect(unwrapToolResultText(envelope)).toBe('total 0\ndrwxr-xr-x@ 2 blamechris staff 64 Jun 14')
+  })
+
+  it('appends stderr after stdout when both are present', () => {
+    const envelope = JSON.stringify({ stdout: 'out line', stderr: 'err line' })
+    expect(unwrapToolResultText(envelope)).toBe('out line\nerr line')
+  })
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/dashboard/src/lib/tool-result-text.test.ts:21]: **FIX** — `6b17f8c3f`. Added the newline-edge-case test directly tied to the `streamsToText` fix:

```ts
it('does not double a trailing newline when stdout already ends in one', () => {
  const envelope = JSON.stringify({ stdout: 'out line\n', stderr: 'err line' })
  // Exactly one newline between the streams — no blank line.
  expect(unwrapToolResultText(envelope)).toBe('out line\nerr line')
})
```

This pins the no-double-newline behaviour so it can't regress.
